pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/PendingAckHandleImpl.java

```diff
@@ -101,36 +103,96 @@
-                pendingAckStoreProvider.newPendingAckStore(persistentSubscription);
-
-        this.pendingAckStoreFuture.thenAccept(pendingAckStore -> {
-            changeToInitializingState();
-            pendingAckStore.replayAsync(this,
-                    ((PersistentTopic) persistentSubscription.getTopic()).getBrokerService()
-                            .getPulsar().getTransactionReplayExecutor());
-        }).exceptionally(e -> {
-            log.error("PendingAckHandleImpl init fail! TopicName : {}, SubName: {}", topicName, subName, e);
-            return null;
+        pendingAckStoreProvider.checkInitializedBefore(persistentSubscription).thenAccept(init -> {
+            if (init) {
+                initPendingAckStore();
+            } else {
+                completeHandleFuture();
+            }
         });
     }
 
+    private void initPendingAckStore() {
+        if (changeToInitializingState()) {
+            synchronized (PendingAckHandleImpl.this) {
+                if (!checkIfClose()) {
+                    this.pendingAckStoreFuture =
+                            pendingAckStoreProvider.newPendingAckStore(persistentSubscription);
+                    this.pendingAckStoreFuture.thenAccept(pendingAckStore -> {
+                        pendingAckStore.replayAsync(this,
+                                ((PersistentTopic) persistentSubscription.getTopic()).getBrokerService()
+                                        .getPulsar().getTransactionReplayExecutor());
+                    }).exceptionally(e -> {
+                        acceptQueue.clear();
+                        changeToErrorState();
+                        log.error("PendingAckHandleImpl init fail! TopicName : {}, SubName: {}", topicName, subName, e);
+                        return null;
+                    });
+                }
+            }
+        }
+    }
+
+    private void addIndividualAcknowledgeMessageRequest(TxnID txnID,
+                                                        List<MutablePair<PositionImpl, Integer>> positions,
+                                                        CompletableFuture<Void> completableFuture) {
+        acceptQueue.add(() -> individualAcknowledgeMessage(txnID, positions, true).thenAccept(v ->
+                completableFuture.complete(null)).exceptionally(e -> {
+            completableFuture.completeExceptionally(e);
+            return null;
+        }));
+    }
+
     @Override
     public CompletableFuture<Void> individualAcknowledgeMessage(TxnID txnID,
-                                                                List<MutablePair<PositionImpl, Integer>> positions) {
+                                                                List<MutablePair<PositionImpl, Integer>> positions,
+                                                                boolean isInCacheRequest) {
+        if (!isInCacheRequest) {
+            if (!checkIfReady()) {
+                synchronized (PendingAckHandleImpl.this) {
```

Comment:
Does this lock is used to make sure the state doesn't be changed?

Reply:
yes